File: lib/rank.js
'use strict'

let _ = require('lodash')
let d = require('debug')('holdem:rank')
let Hand = require('./hand')


var HandRank = class handRank {

  constructor (hands) {
    this.handRankPoints = {
      'Royal Flush': 10,
      'Straight Flush': 9,
      '4 of a Kind': 8,
      'Full House': 7,
      'Flush': 6,
      'Straight': 5,
      '3 of a Kind': 4,
      '2 Pair': 3,
      '1 Pair': 2,
      'High Card': 1
    }
    this.cardMap = ['2','3','4','5','6','7','8','9','T','J','Q','K','A']
    this.hands = []
    if (typeof hands !== 'object' || hands.length < 2) {
      throw new Error('HandRank expects at least two hands')
    }

    this.hands = hands.map((hand) => {
      return new Hand(hand)
    })
  }

  //high card, pair, 2 pairs, 3 of a kind, straight, flush, full house, 4 of a kind, straight flush, royal flush

  rank () {
    this.hands.map((h) => {
      //h.highCard = this._highCard(h)
      this._setRank(h, 'High Card')
      if(this._isRoyalFlush(h)) {
        this._setRank(h, 'Royal Flush')
      } else if(this._isStraightFlush(h)) {
        this._setRank(h, 'Straight Flush')
      } else if(this._isFourOfaKind(h)) {
        this._setRank(h, '4 of a Kind')
      } else if(this._isFullHouse(h)) {
        this._setRank(h, 'Full House')
      } else if (this._isFlush(h)) {
        this._setRank(h, 'Flush')
      } else if (this._isStraight(h)) {
        this._setRank(h, 'Straight')
      } else if (this._isThreeOfaKind(h)) {
        this._setRank(h, '3 of a Kind')
      } else if (this._isTwoPair(h)) {
        this._setRank(h, '2 Pair')
      } else if (this._isOnePair(h)) {
        this._setRank(h, '1 Pair')
      }
    })

    return this.hands
  }

  _setRank(hand, rankName) {
    hand.rank = this.handRankPoints[rankName]
  }

  _isRoyalFlush (hand) {
    if (this._isFlush(hand)) {
      return (
          hand.value[0] === 14 &&
          hand.value[1] === 13 &&
          hand.value[2] === 12 &&
          hand.value[3] === 11 &&
          hand.value[4] === 10
      )
    } 
    return false
  }

  _isStraightFlush (hand) {
    if (this._isFlush(hand) && this._isStraight(hand)) {
      return true
    }
    return false
  }

  _isFlush (hand) {
    var match = hand.suits.match(/CCCCC|DDDDD|HHHHH|SSSSS/g)
    return match != null
  }

  _isStraight (hand) {
    var straight = true, i
    for (i = 0; i < hand.value.length - 1; i++) {
      if(hand.value[i] !== hand.value[i+1] + 1) {
        straight = false
      }
    }

    return straight
  }

  _isOnePair (hand) {
    var pairCount = 0
    for (var card in hand.hash) {
      if(hand.hash[card] === 2) {
        pairCount++
      }
    }
    return pairCount === 1
  }

  _isTwoPair (hand) {
    var pairCount = 0
    for (var card in hand.hash) {
      if(hand.hash[card] === 2) {
        pairCount++
      }
    }
    return pairCount === 2
  }

  _isThreeOfaKind (hand) {
    var threes = 0

    for (var card in hand.hash) {
      if(hand.hash[card] === 3) {
        threes++
      }
    }

    return threes === 1
  }

  _isFullHouse (hand) {
    return (this._isThreeOfaKind(hand) && this._isOnePair(hand))
  }

  _isFourOfaKind (hand) {
    var fours = 0
    for (var card in hand.hash) {
      if(hand.hash[card] === 4) {
        fours++
      }
    }

    return fours === 1
  }

  /*_highCard (hand) {
    for(var i = hand.hash.length; i > 0; i--) {
      if(hand.hash[i] > 0) {
        return this.cardMap[i]
      }
    }
  }*/
}

module.exports = HandRank

File: java/holdem/src/holdem/Rank.java
package holdem;

import java.util.*;

/**
 *
 * @author nilotpal
 */
public class Rank {
    private HashMap<Integer, String> handRanks = new HashMap();
    
    public Rank() {
        buildhandRanks();
    }
    
    private void buildhandRanks() {
        handRanks.put(10, "Royal Flush");
        handRanks.put(9, "Straight Flush");
        handRanks.put(8, "Four of a Kind");
        handRanks.put(7, "Full House");
        handRanks.put(6, "Flush");
        handRanks.put(5, "Straight");
        handRanks.put(4, "Three of a Kind");
        handRanks.put(3, "Two Pair");
        handRanks.put(2, "One Pair");
        handRanks.put(1, "High Card");
    }
    
    public Hand analyze(Hand hand) {
        this.setRank(hand, 1);
        if(this.isRoyalFlush(hand)) {
            this.setRank(hand, 10);
        } else if(this.isStraightFlush(hand)) {
            this.setRank(hand, 9);
        } else if(this.isFourOfaKind(hand)) {
            this.setRank(hand, 8);
        } else if(this.isFullHouse(hand)) {
            this.setRank(hand, 7);
        } else if(this.isFlush(hand)) {
            this.setRank(hand, 6);
        } else if(this.isStraight(hand)) {
            this.setRank(hand, 5);
        } else if(this.isThreeOfaKind(hand)) {
            this.setRank(hand, 4);
        } else if(this.isTwoPair(hand)) {
            this.setRank(hand, 3);
        } else if(this.isOnePair(hand)) {
            this.setRank(hand, 2);
        }
        return hand;
    }
    
    private void setRank (Hand hand, int rank) {
        hand.setRank(handRanks.get(rank), rank);
    }
    
    private boolean isFlush(Hand hand) {
        return (
                hand.suits().equals("CCCCC") || 
                hand.suits().equals("DDDDD") || 
                hand.suits().equals("HHHHH") ||
                hand.suits().equals("SSSSS")
        );
    }
    
    private boolean isRoyalFlush(Hand hand) {
        //System.out.println(Arrays.toString(hand.values()));
        if(this.isFlush(hand)) {
            return (
                    hand.values()[0] == 14 &&
                    hand.values()[1] == 13 &&
                    hand.values()[2] == 12 &&
                    hand.values()[3] == 11 &&
                    hand.values()[4] == 10
            );
        }
        
        return false;
    }
    
    private boolean isStraightFlush(Hand hand) {
        return (this.isFlush(hand) && this.isStraight(hand));
    }
    
    private boolean isStraight(Hand hand) {
        Integer[] values = hand.values();
        boolean straight = true;
        int i;
        
        for(i = 0; i < values.length - 1; i++) {
            if(values[i] != values[i+1] + 1) {
                straight = false;
            }
        }
        
        return straight;
    }
    
    private boolean isOnePair(Hand hand) {
        int pairCount = 0;
        for(Integer value : hand.hash().values()) {
            if(value == 2) {
                pairCount++;
            }
        }
        return pairCount == 1;
    }
    
    private boolean isTwoPair(Hand hand) {
        int pairCount = 0;
        for(Integer value : hand.hash().values()) {
            if(value == 2) {
                pairCount++;
            }
        }
        return pairCount == 2;
    }
    
    private boolean isThreeOfaKind(Hand hand) {
        int threes = 0;
        for(Integer value : hand.hash().values()) {
            if(value == 3) {
                threes++;
            }
        }
        
        return threes == 1;
    }
    
    private boolean isFourOfaKind(Hand hand) {
        int fours = 0;
        for(Integer value : hand.hash().values()) {
            if(value == 4) {
                fours++;
            }
        }
        return fours == 1;
    }
    
    private boolean isFullHouse(Hand hand) {
        return (this.isThreeOfaKind(hand) && this.isOnePair(hand));
    }
}
